From 51da0dcf515bd63d44dd4c479147810b4f30a7f5 Mon Sep 17 00:00:00 2001
From: Ulrich Drepper <drepper@redhat.com>
Date: Thu, 2 Jul 2009 03:30:55 -0700
Subject: [PATCH 048/100] Fix possible race when freeing object in fast bin list.
 (cherry picked from commit 6cbbaa50aac809ad6e0692247876c82d58e466bf)

---
 ChangeLog       |    5 +++++
 malloc/malloc.c |    1 +
 2 files changed, 6 insertions(+), 0 deletions(-)

2009-07-02  Ulrich Drepper  <drepper@redhat.com>

	* malloc/malloc.c [ATOMIC_FASTBINS] (_int_free): Add full barrier when
	adding to fast bin list.

diff --git a/malloc/malloc.c b/malloc/malloc.c
index 0b9face..703dcb5 100644
--- a/malloc/malloc.c
+++ b/malloc/malloc.c
@@ -4802,6 +4802,7 @@ _int_free(mstate av, mchunkptr p)
 	    goto errout;
 	  }
 	p->fd = fd = old;
+	atomic_full_barrier ();
       }
     while ((old = catomic_compare_and_exchange_val_acq (fb, p, fd)) != fd);
 #else
-- 
1.6.5.2

